[17.0][MIG] sale_blanket_order

niboo-jva opened the pull request.

andreparames committed:
[10.0][ADD] sale_blanket_orders

andreparames committed:
sale_blanket_order: don't copy name nor confirm state

When duplicating a confirmed blanket order, the new copy shouldn't
keep the state nor the sequence number (name).

andreparames committed:
sale_blanket_order: update after review

andreparames committed:
sale_blanket_order: Re-calculate price based on qty

andreparames committed:
sale_blanket_order: make prohibition of adding new lines optional

rousseldenis committed:
[FIX] sale_blanket_order: Ondelete cascade on blanket lines

AdriaGForgeFlow committed:
[11.0][MIG] Migrate module sale_blanket_order

marylla committed:
Added translation using Weblate (German)

marylla committed:
Translated using Weblate (German)

Currently translated at 7.3% (9 of 123 strings)

Translation: sale-workflow-11.0/sale-workflow-11.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-11-0/sale-workflow-11-0-sale_blanket_order/de/

tafaRU committed:
[MIG] sale_blanket_order: Migration to 12.0

tafaRU committed:
sale_blanket_order: fix the following use case scenario.

Steps to reproduce:

* create and confirm a blanket order (BO) with (product A, qty 30) and (product B, qty 20)
* from the BO create a SO with (product A, qty 10) and (product B, qty 10)
* from the BO create a SO with (product A, qty 20) and (product B, qty 0)
* from the BO create another SO with (product B, qty 10)

Current behavior:

It raises the exception "The sale has already been completed.".

Expected behavior:

No exception is raised.

tafaRU committed:
sale_blanket_order: correctly display images in both README and index

tafaRU committed:
sale_blanket_order: fix my email address format in CONTRIBUTORS.rst

Galoperd committed:
Added translation using Weblate (Spanish)

Galoperd committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (154 of 154 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/es/

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

AndreVolksdorf committed:
Translated using Weblate (German)

Currently translated at 7.1% (11 of 154 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/de/

AndreVolksdorf committed:
Translated using Weblate (German)

Currently translated at 53.9% (83 of 154 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/de/

AndreVolksdorf committed:
Translated using Weblate (German)

Currently translated at 100.0% (154 of 154 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/de/

AndreVolksdorf committed:
Translated using Weblate (German)

Currently translated at 100.0% (154 of 154 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/de/

chusamo committed:
[MIG]sale_blanker_order: Migration to 13.0

LoisRForgeFlow committed:
[IMP] sale_blanket_order: use default terms and conditions.

exactly like in sales orders.

LoisRForgeFlow committed:
[FIX] sale_blanket_order: multi-company rules were not migrated properly.

LoisRForgeFlow committed:
[IMP] sale_blanket_order: analytic acc and origin

* Allow to select an analitic account in the blanket order. It
  will be passed on to the SO.
* Allow to select analytic tags on blanket order lines. They will
  be passed to the SO lines.
* Source document was not being filled in SO created from a
  blanket order.

tafaRU committed:
[IMP] sale_blanket_order: black, isort, prettier

tafaRU committed:
[MIG] sale_blanket_order: Migration to 14.0

tafaRU committed:
[FIX] sale_blanket_order: do not use supplier_taxes_id

tafaRU committed:
[FIX] sale_blanket_order: do not recompute tax_id on SO line if already set

Otherwise use case same as that one written in https://github.com/OCA/sale-workflow/blob/14.0/sale_shipping_info_helper/tests/test_sale_shipping_info_helper.py#L22 fails
See https://app.travis-ci.com/github/OCA/sale-workflow/jobs/555432793#L3103-L3107 as reference

oca-travis committed:
[UPD] Update sale_blanket_order.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
sale_blanket_order 14.0.1.0.1

JasminSForgeFlow committed:
[IMP] sale_blanket_order: black, isort, prettier

JasminSForgeFlow committed:
[MIG] sale_blanket_order: Migration to 15.0

BernatPForgeFlow committed:
[IMP] sale_blanket_order: Add prepare SO values hook

Commit:
[UPD] Update sale_blanket_order.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

Commit:
Added translation using Weblate (French (France))

Commit:
Translated using Weblate (French (France))

Currently translated at 3.7% (6 of 161 strings)

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_blanket_order/fr_FR/

ClementChaumienneC2C committed:
Translated using Weblate (French (France))

Currently translated at 99.3% (160 of 161 strings)

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_blanket_order/fr_FR/

leemannd committed:
[FIX] sale_blanket_order: total in pdf report not visible

ClementChaumienneC2C committed:
Translated using Weblate (French (France))

Currently translated at 100.0% (161 of 161 strings)

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_blanket_order/fr_FR/

Commit:
[UPD] Update sale_blanket_order.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_blanket_order/

OCA-git-bot committed:
sale_blanket_order 15.0.1.0.1

ClementChaumienneC2C committed:
Translated using Weblate (French (France))

Currently translated at 100.0% (162 of 162 strings)

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_blanket_order/fr_FR/

santostelmo committed:
[IMP] Add section note to sale blanket order

OCA-git-bot committed:
sale_blanket_order 15.0.1.1.0

Commit:
[UPD] Update sale_blanket_order.pot

OCA-git-bot committed:
sale_blanket_order 15.0.1.2.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_blanket_order/

ClementChaumienneC2C committed:
Translated using Weblate (French (France))

Currently translated at 100.0% (172 of 172 strings)

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_blanket_order/fr_FR/

leemannd committed:
Added translation using Weblate (French)

leemannd committed:
Translated using Weblate (French)

Currently translated at 55.2% (95 of 172 strings)

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_blanket_order/fr/

ClementChaumienneC2C committed:
Translated using Weblate (French)

Currently translated at 100.0% (172 of 172 strings)

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_blanket_order/fr/

nguyenminhchien committed:
[16.0][MIG] sale_blanket_order: Migration to 16.0

Commit:
[UPD] Update sale_blanket_order.pot

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (170 of 170 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_blanket_order/it/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (172 of 172 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_blanket_order/es/

pedrobaeza commented:
Thanks for the contribution.

Please preserve commit history following technical method explained in https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0.

If the jump is between several versions, you have to modify the source branch in the main command to accommodate it to this circumstance.

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (170 of 170 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_blanket_order/it/

niboo-jva committed:
[IMP] sale_blanket_order: pre-commit auto fixes

niboo-jva committed:
[MIG] sale_blanket_order: Migration to 17.0

rousseldenis commented:
/ocabot migration sale_blanket_order

rousseldenis reviewed (approved):
Code review

Deriman-Alonso reviewed (changes requested):
Hi, on tree views invisible must be replaced by column_invisible, Thanks

chrisb-c01 commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -564,34 +556,44 @@ def _get_real_price_currency(self, product, rule_id, qty, uom, pricelist_id):
         return product[field_name] * uom_factor * cur_factor, currency_id.id
 
     def _get_display_price(self, product):
-        # Copied and adapted from the sale module
```
Please leave this comment in for traceability purposes.

chrisb-c01 commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -564,34 +556,44 @@ def _get_real_price_currency(self, product, rule_id, qty, uom, pricelist_id):
+            product=self.product_id,
+            quantity=self.original_uom_qty or 1.0,
+            uom=self.product_uom,
+            date=self.order_id.validity_date,
```
I don't think you should use `validity_date` here. The `validity_date` defines the date until which the blanket order is valid. The date entered in the `_compute_price()` function should be the 'date to use for price computation and currency conversions'. I see `fields.Date.today()` is used in earlier versions to compute the price.

chrisb-c01 commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -564,34 +556,44 @@ def _get_real_price_currency(self, product, rule_id, qty, uom, pricelist_id):
+        return max(base_price, pricelist_price)
+
+    def _get_pricelist_price_before_discount(self):
+        """Compute the price used as base for the pricelist price computation.
```
This is a copy & paste from the sale module. Mentioning it would be helpful for future migrations.

E.g.: `# Copied and adapted from the sale module`

chrisb-c01 commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -564,34 +556,44 @@ def _get_real_price_currency(self, product, rule_id, qty, uom, pricelist_id):
+            product=self.product_id,
+            quantity=self.product_uom_qty or 1.0,
+            uom=self.product_uom,
+            date=self.order_id.date_order,
```
`order_id` has no field `date_order`. Copy & paste mistake?

chrisb-c01 commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -564,34 +556,44 @@ def _get_real_price_currency(self, product, rule_id, qty, uom, pricelist_id):
 
     def _get_display_price(self, product):
-        # Copied and adapted from the sale module
         self.ensure_one()
```
Why not include the `self.product_id.ensure_one()` here as well, similar to the function in the sale module?

chrisb-c01 commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -656,6 +658,23 @@ def _compute_quantities(self):
 
+    @api.depends("product_id", "product_uom", "original_uom_qty")
+    def _compute_pricelist_item_id(self):
+        for line in self:
```
Please add `# Copied and adapted from the sale module`

chrisb-c01 commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -656,6 +658,23 @@ def _compute_quantities(self):
+                    line.product_id,
+                    quantity=line.original_uom_qty or 1.0,
+                    uom=line.product_uom,
+                    date=line.order_id.validity_date,
```
`validity_date` shouldn't be used here I think (see my other comment). Perhaps best to use `fields.Date.today()`.

chrisb-c01 commented on sale_blanket_order/models/sale_orders.py:
```diff
@@ -61,14 +61,14 @@ def _get_assigned_bo_line(self, bo_lines):
         date_planned = date.today()
         date_delta = timedelta(days=365)
-        for line in bo_lines.filtered(lambda l: l.date_schedule):
+        for line in bo_lines.filtered(lambda bo_line: bo_line.date_schedule):
```
Is this change really needed?

chrisb-c01 replied:
> Is this change really needed?

Apparently this is needed because of a linting error:

> E741 Ambiguous variable name: `l`

chrisb-c01 commented on sale_blanket_order/views/sale_blanket_order_views.xml:
```diff
@@ -37,25 +37,25 @@
                     <button
                         name="set_to_draft"
-                        states="expired,done"
+                        invisible="state != ['expired', 'done']"
```
Do you mean `state not in ['expired', 'done']`?

```
>>> state = 'ready'
>>> state != ['expired', 'done']
True
>>> state not in ['expired', 'done']
True
>>> state = 'done'
>>> state != ['expired', 'done']
True
>>> state not in ['expired', 'done']
False
```

chrisb-c01 commented on sale_blanket_order/views/sale_blanket_order_views.xml:
```diff
@@ -146,10 +143,8 @@
-                                            'invisible': [('display_type', '=', True)],
-                                        }"
+                                        required="display_type == False"
+                                        invisible="display_type == True"
```
Please use `column_invisible` see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

chrisb-c01 commented on sale_blanket_order/views/sale_blanket_order_views.xml:
```diff
@@ -181,10 +176,8 @@
-                                            'invisible': [('display_type', '=', True)],
-                                        }"
+                                        required="display_type == False"
+                                        invisible="display_type == True"
```
Please use `column_invisible` see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

chrisb-c01 commented on sale_blanket_order/views/sale_blanket_order_views.xml:
```diff
@@ -198,10 +191,8 @@
-                                            'invisible': [('display_type', '=', True)],
-                                        }"
+                                        required="display_type == False"
+                                        invisible="display_type == True"
```
Please use `column_invisible` see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

chrisb-c01 commented on sale_blanket_order/views/sale_blanket_order_views.xml:
```diff
@@ -214,17 +205,13 @@
-                                            'invisible': [('display_type', '=', True)],
-                                        }"
+                                        required="display_type == False"
+                                        invisible="display_type == True"
```
Please use `column_invisible` see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

chrisb-c01 commented on sale_blanket_order/views/sale_blanket_order_views.xml:
```diff
@@ -214,17 +205,13 @@
-                                        attrs="{
-                                            'invisible': [('display_type', '=', True)],
-                                        }"
+                                        invisible="display_type == True"
```
Please use `column_invisible` see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

chrisb-c01 commented on sale_blanket_order/views/sale_blanket_order_views.xml:
```diff
@@ -284,7 +271,7 @@
                                         <field
                                             name="analytic_account_id"
-                                            attrs="{'readonly': [('sale_count','!=',0),('state','!=','draft')]}"
+                                            readonly="sale_count != 0 or state != 'draft'"
```
Shouldn't the `or` be and `and` as no or-operator (`'|'`) is present in de `attrs` domain?

chrisb-c01 commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -483,6 +472,9 @@ def _compute_amount(self):
+        comodel_name="product.pricelist.item", compute="_compute_pricelist_item_id"
+    )
 
     def name_get(self):
```
This should be updated as per https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0 

> If you are overriding name_get method, you should now override _compute_display_name instead, adding the possible new fields in the depends of the method. See https://github.com/odoo/odoo/pull/122085 for more details.

chrisb-c01 commented:
Hi @jvandri thanks for your effort on getting this to v17. Will you have some time to apply the fixes so that we can get this migrated?

chrisb-c01 commented on sale_blanket_order/i18n/it.po:
```diff
@@ -6,7 +6,7 @@ msgid ""
 "Project-Id-Version: Odoo Server 16.0\n"
 "Report-Msgid-Bugs-To: \n"
-"PO-Revision-Date: 2024-03-15 16:34+0000\n"
+"PO-Revision-Date: 2024-01-25 15:34+0000\n"
```
Was this intentional (to go back in time)?

chrisb-c01 commented:
I've created a new PR based on this one and applied my changes. https://github.com/OCA/sale-workflow/pull/3255

rousseldenis commented:
@jvandri We switch to #3255
